realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -85,45 +91,64 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
 
     // Realm path will be used as the key to store different RealmCaches. Different Realm configurations with same path
     // are not allowed and an exception will be thrown when trying to add it to the cache map.
-    private static final Map<String, RealmCache> cachesMap = new HashMap<>();
+    // A weak ref is used to hold the RealmCache instance. The weak ref entry will be cleared if and only if there
+    // is no Realm instance holding a strong ref to it and there is no Realm instance associated it is BEING created.
+    private static final Map<String, WeakReference<RealmCache>> cachesMap =
+            new HashMap<String, WeakReference<RealmCache>>();
 
     private static final String DIFFERENT_KEY_MESSAGE = "Wrong key used to decrypt Realm.";
     private static final String WRONG_REALM_CLASS_MESSAGE = "The type of Realm class must be Realm or DynamicRealm.";
 
-    private RealmCache(RealmConfiguration config) {
-        configuration = config;
+    private RealmCache() {
         refAndCountMap = new EnumMap<>(RealmCacheType.class);
         for (RealmCacheType type : RealmCacheType.values()) {
             refAndCountMap.put(type, new RefAndCount());
         }
     }
 
+    private static synchronized RealmCache getCache(String realmPath, boolean createIfNotExist) {
```

Comment:
Why is this synchronized?  It looks like it might is because there is a concern about multi-threaded access to `cacheMap`.  If that is the case, then you *must* synchronize all other access to `cacheMap` on the same lock... and I suggest that you might consider using a lock other than the Class object.